provider/mesos.go

```diff
@@ -0,0 +1,449 @@
+	}
+	return "false"
+}
+
+func (provider *Mesos) getPriority(task state.Task) string {
+	if priority, err := provider.getLabel(task, "traefik.frontend.priority"); err == nil {
+		return priority
+	}
+	return "0"
+}
+
+func (provider *Mesos) getEntryPoints(task state.Task) []string {
+	if entryPoints, err := provider.getLabel(task, "traefik.frontend.entryPoints"); err == nil {
+		return strings.Split(entryPoints, ",")
+	}
+	return []string{}
+}
+
+// getFrontendRule returns the frontend rule for the specified application, using
+// it's label. It returns a default one (Host) if the label is not present.
+func (provider *Mesos) getFrontendRule(task state.Task) string {
+	// ⚠⚠⚠⚠⚠⚠⚠⚠⚠⚠⚠⚠⚠⚠⚠⚠⚠⚠⚠⚠⚠⚠⚠⚠⚠⚠⚠⚠⚠⚠⚠⚠⚠⚠⚠⚠⚠⚠
+	// TODO: backwards compatibility with DEPRECATED rule.Value
+	if value, err := provider.getLabel(task, "traefik.frontend.value"); err == nil {
+		log.Warnf("Label traefik.frontend.value=%s is DEPRECATED, please refer to the rule label: https://github.com/containous/traefik/blob/master/docs/index.md#mesos", value)
+		rule, _ := provider.getLabel(task, "traefik.frontend.rule")
+		return rule + ":" + value
+	}
+	// ⚠⚠⚠⚠⚠⚠⚠⚠⚠⚠⚠⚠⚠⚠⚠⚠⚠⚠⚠⚠⚠⚠⚠⚠⚠⚠⚠⚠⚠⚠⚠⚠⚠⚠⚠⚠⚠⚠
+	if label, err := provider.getLabel(task, "traefik.frontend.rule"); err == nil {
+		return label
+	}
+	return "Host:" + strings.ToLower(strings.Replace(provider.getSubDomain(task.DiscoveryInfo.Name), "_", "-", -1)) + "." + provider.Domain
+}
+
+func (provider *Mesos) getBackend(task state.Task, applications []state.Task) string {
+	application, errApp := getMesos(task, applications)
+	if errApp != nil {
+		log.Errorf("Unable to get mesos application from task %s", task.DiscoveryInfo.Name)
+		return ""
+	}
+	return provider.getFrontendBackend(application)
+}
+
+func (provider *Mesos) getFrontendBackend(task state.Task) string {
+	if label, err := provider.getLabel(task, "traefik.backend"); err == nil {
+		return label
+	}
+	return "-" + strings.Replace(strings.Replace(task.DiscoveryInfo.Name, ".", "-", -1), "_", "-", -1)
+}
+
+func (provider *Mesos) getHost(task state.Task) string {
+	return task.IP(strings.Split(provider.IPSources, ",")...)
+}
+
+func (provider *Mesos) getID(task state.Task) string {
+	return strings.Replace(strings.Replace(strings.Replace(task.ID, ".", "-", -1), ":", "-", -1), "_", "-", -1)
+}
+
+func (provider *Mesos) getFrontEndName(task state.Task) string {
+	return strings.Replace(strings.Replace(strings.Replace(strings.Replace(task.ID, ".", "-", -1), ":", "-", -1), "_", "-", -1), "/", "-", -1)
+}
+
+func detectMasters(zk string, masters []string) <-chan []string {
+	changed := make(chan []string, 1)
+	if zk != "" {
+		log.Debugf("Starting master detector for ZK ", zk)
+		if md, err := detector.New(zk); err != nil {
+			log.Fatalf("failed to create master detector: %v", err)
+		} else if err := md.Detect(detect.NewMasters(masters, changed)); err != nil {
+			log.Fatalf("failed to initialize master detector: %v", err)
+		}
+	} else {
+		changed <- masters
```

Comment:
I don't understand this point ...
Maybe you want to point the fact that if zk is not defined, we put masters in the variable "changed" but nothing validate that "master" is defined.